lib/ansible/modules/cloud/amazon/lambda_policy.py

```diff
@@ -0,0 +1,442 @@
+
+    :param module:
+    :param client:
+    :return:
+    """
+
+    policy = dict()
+    sid = module.params['statement_id']
+
+    # set API parameters
+    api_params = set_api_params(module, ('function_name', ))
+    qualifier = get_qualifier(module)
+    if qualifier:
+        api_params.update(Qualifier=qualifier)
+
+    policy_results = None
+    # check if function policy exists
+    try:
+        policy_results = client.get_policy(**api_params)
+    except ClientError as e:
+        try:
+            if e.response['Error']['Code'] == 'ResourceNotFoundException':
+                return {}
+        except:  # catches ClientErrors without response, e.g. fail before connect
+            pass
+        module.fail_json_aws(e, msg="retrieving function policy")
+    except Exception as e:
+        module.fail_json_aws(e, msg="retrieving function policy")
+
+    # get_policy returns a JSON string so must convert to dict before reassigning to its key
+    policy = json.loads(policy_results.get('Policy', '{}'))
+    return extract_statement(policy, sid)
+
+
+def add_policy_permission(module, client):
+    """
+    Adds a permission statement to the policy.
+
+    :param module:
+    :param aws:
+    :return:
+    """
+
+    changed = False
+
+    # set API parameters
+    params = (
+        'function_name',
+        'statement_id',
+        'action',
+        'principal',
+        'source_arn',
+        'source_account',
+        'event_source_token')
+    api_params = set_api_params(module, params)
+    qualifier = get_qualifier(module)
+    if qualifier:
+        api_params.update(Qualifier=qualifier)
+
+    if not module.check_mode:
+        try:
+            client.add_permission(**api_params)
+        except Exception as e:
+            module.fail_json_aws(e, msg="adding permission to policy")
+        changed = True
+
+    return changed
+
+
+def remove_policy_permission(module, client):
+    """
+    Removed a permission statement from the policy.
+
+    :param module:
+    :param aws:
+    :return:
+    """
+
+    changed = False
+
+    # set API parameters
+    api_params = set_api_params(module, ('function_name', 'statement_id'))
+    qualifier = get_qualifier(module)
+    if qualifier:
+        api_params.update(Qualifier=qualifier)
+
+    try:
+        if not module.check_mode:
+            client.remove_permission(**api_params)
+            changed = True
+    except Exception as e:
+        module.fail_json_aws(e, msg="removing permission from policy")
+
+    return changed
+
+
+def manage_state(module, lambda_client):
+    changed = False
+    current_state = 'absent'
+    state = module.params['state']
+    action_taken = 'none'
+
+    # check if the policy exists
+    current_policy_statement = get_policy_statement(module, lambda_client)
+    if current_policy_statement:
+        current_state = 'present'
+
+    if state == 'present':
+        if current_state == 'present':
```

Comment:
`if current_state == 'present' and not policy_equal(module, current_policy_statement):`